### PR TITLE
[Relay] GNF

### PR DESCRIPTION
**Pull request by MarisaKirisame**
This PR will do the exact reverse of ANF conversion - it take any code using let, and will attempt to remove the let by reusing the same expr.
The only nontrivial case is recursion - we keep let in such case to avoid forming loop.

### CODE REVIEW

**Comment by MarisaKirisame** [comment]
@srkreddy1238 @vinx13 @yzhliu can you guys help review? Thx!

**Comment by slyubomirsky** [comment]
If recursion is a corner case, can you have a recursive test case?

File: include/tvm/relay/pass.h [hunk 1]
```diff
@@ -296,6 +296,17 @@ struct StructuralHash {
...
+/*! \brief Remove let binding and directly share via pointer instead.
+ *
+ * It will remove all let binding,
+ * and turn all of the variable bound by let into direct pointer reference.
+ *
+ * \param e the expression.
+ *
+ * \return the expression in graph normal form.
+ */
+Expr ToGNF(const Expr& e);
```
Comment (tqchen) [inline include/tvm/relay/pass.h]: GNF is not a common term, perhaps ToGraphNormalForm?

Comment (Ravenwater) [reply include/tvm/relay/pass.h]: +1 for ToGraphNormalForm: please use descriptive names so that others that come from a different angle can read the source code.

**Comment by MarisaKirisame** [comment]
@slyubomirsky recursion need pass manager. i will add it once it get in.

**Comment by slyubomirsky** [comment]
What would the pass manager be needed to do?

**Comment by MarisaKirisame** [comment]
pass manager will apply a pass into a golbalvar. somehow fuse_ops cant handle recursion without it.

File: src/relay/pass/to_gnf.cc [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/*!
```
Comment (tqchen) [inline src/relay/pass/to_gnf.cc]: rename to to_graph_normal_form

File: src/relay/pass/to_gnf.cc [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ */
+#include <tvm/relay/pass.h>
+#include <tvm/relay/expr_functor.h>
+#include "let_list.h"
+
+namespace tvm {
+namespace relay {
+
+class GNF : public ExprMutator {
+  std::unordered_map<Var, Expr, NodeHash, NodeEqual> var_map_;
```
Comment (tqchen) [inline src/relay/pass/to_gnf.cc]: mark field as private

File: src/relay/pass/to_gnf.cc [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  Expr VisitExpr_(const VarNode* vn) override {
+    Var v = GetRef<Var>(vn);
+    return var_map_.count(v) == 0 ? v : var_map_.at(v);
+  }
+
+  static bool UseVar(const Var& v, const Expr& e) {
```
Comment (tqchen) [inline src/relay/pass/to_gnf.cc]: Let us make a quick util function for UseVar, since FreeVars is not the best way to do so

File: tests/python/relay/test_to_gnf.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import numpy as np
+import tvm
+from tvm import relay
+from tvm.relay.ir_pass import to_graph_normal_form, alpha_equal
+from tvm.relay import op, create_executor
+from tvm.relay.backend.interpreter import Value, TupleValue
+
+
+def check_eval(expr, args, expected_result, mod=None, rtol=1e-07):
```
Comment (tqchen) [inline tests/python/relay/test_to_gnf.py]: write test cases that does ANF -> GNF roundtrip

File: tests/python/relay/test_to_gnf.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import numpy as np
```
Comment (tqchen) [inline tests/python/relay/test_to_gnf.py]: test_to_graph_normal_form.py

File: tests/python/relay/test_to_gnf.py (hunk 5 above)
Comment (tqchen) [reply tests/python/relay/test_to_gnf.py]: Get an existing neural network, run ToANF, and then ToGraphNormalForm and check consistency

**Comment by tqchen** [comment]
Followup comment: to make API consistent, maybe we should rename to_anf to to_a_normal_form

File: tests/python/relay/test_to_gnf.py (hunk 5 above)
Comment (MarisaKirisame) [reply tests/python/relay/test_to_gnf.py]: is it necessary to do an nn? i get a simpler test working

**Comment by MarisaKirisame** [comment]
@tqchen can you review? sorry for the inconvenience.
